store/sqlstore/upgrade.go

```diff
@@ -965,6 +965,11 @@ func upgradeDatabaseToVersion532(sqlStore *SqlStore) {
 	// allow 10 files per post
 	sqlStore.AlterColumnTypeIfExists("Posts", "FileIds", "text", "varchar(300)")
 	sqlStore.CreateColumnIfNotExistsNoDefault("Channels", "Shared", "tinyint(1)", "boolean")
+	sqlStore.CreateColumnIfNotExistsNoDefault("Reactions", "UpdateAt", "bigint", "bigint")
+	sqlStore.CreateColumnIfNotExistsNoDefault("Reactions", "DeleteAt", "bigint", "bigint")
```

Comment:
The reactions table has a very important multi-column index which is _very_ sensitive to performance. If we want to add these columns, we need to verify performance and possibly add new indexes/update existing ones.

Customers have extremely large reactions tables, so we might also want to run some simulations on a 10M reactions table and see how long these column additions take.

Basically, just some due diligence on index performance, and verifying that there are no regressions.

Reply:
To second what @agnivade said, would be awesome if we can have some explains on a moderately large reactions table.

Reply:
This being handled via the Perf Review tag that Claudio is helping with.

Reply:
I've run the upgrade steps on both db engines (running on `db.r5.4xlarge`) using a rather big reactions table (~8.5M entries). Here are the results:

#### PostgreSQL

```
=> ALTER TABLE Reactions ADD COLUMN UpdateAt bigint;
ALTER TABLE
Time: 3.851 ms
=> ALTER TABLE Reactions ADD COLUMN DeleteAt bigint;
ALTER TABLE
Time: 3.670 ms
=> EXPLAIN ANALYZE UPDATE Reactions SET UpdateAt=CreateAt, DeleteAt=0 WHERE DeleteAt IS NULL;
                                                         QUERY PLAN                                                         
----------------------------------------------------------------------------------------------------------------------------
 Update on reactions  (cost=0.00..191756.34 rows=42907 width=88) (actual time=137371.651..137371.651 rows=0 loops=1)
   ->  Seq Scan on reactions  (cost=0.00..191756.34 rows=42907 width=88) (actual time=0.013..2539.940 rows=8581334 loops=1)
         Filter: (deleteat IS NULL)
 Planning Time: 0.186 ms
 Execution Time: 137371.675 ms
(5 rows)

Time: 137461.137 ms (02:17.461)
```

#### MySQL

```
> ALTER TABLE Reactions ADD COLUMN UpdateAt bigint;
Query OK, 0 rows affected (12.75 sec)
Records: 0  Duplicates: 0  Warnings: 0

> ALTER TABLE Reactions ADD COLUMN DeleteAt bigint;
Query OK, 0 rows affected (13.05 sec)
Records: 0  Duplicates: 0  Warnings: 0

> UPDATE Reactions SET UpdateAt=CreateAt, DeleteAt=0 WHERE DeleteAt IS NULL;
Query OK, 8581334 rows affected (4 min 30.41 sec)
Rows matched: 8581334  Changed: 8581334  Warnings: 0
```

As expected the `UPDATE` query can take a significant amount of time. I guess we need to decide whether having the server stuck for a few minutes is considered acceptable (although this time could significantly change between different db setups). If not, with a bit more logic I think we should be able to handle this on the application side, essentially checking for zero `UpdateAt`/`DeleteAt` values (through the use of `COALESCE`).

Reply:
Also ran a comparative load-test on the same dataset which did **not** reveal any performance issues/regressions:

```
==================================================
Comparison result:
Report: report_0_postgresql_unbounded.md
Grafana Dashboard: http://54.162.212.152:3000/d/6ZPSLWBGz/comparison-0-postgresql-unbounded
5.30.1:
  DB Engine: postgresql
  Supported Users: 2611
  Errors: 3094
MM-31875:
  Type: unbounded
  DB Engine: postgresql
  Supported Users: 2698
  Errors: 2983
==================================================

==================================================
Comparison result:
Grafana Dashboard: http://52.200.68.215:3000/d/_0NNYZBGz/comparison-0-mysql-unbounded
5.30.1:
  Type: unbounded
  DB Engine: mysql
  Supported Users: 2557
  Errors: 2896
MM-31875:
  Type: unbounded
  DB Engine: mysql
  Supported Users: 2677
  Errors: 2520
==================================================
```

As far as db indexes I am going to suggest **not** adding/modifying anything. The existing one on the multi-column primary key is working extremely well. 

I've benchmarked a possible addition of a `Reactions.PostId, Reactions.DeleteAt` index which did not show any performance improvement. This is somewhat expected given that the costliest part is getting reactions for a given `postId` after which the number of entries is already drastically reduced. 

Therefore, unless we expect degenerate cases of tens of thousands of reactions per post, I'd say we should be okay with what we already have.

```
                       base time/op  with_added_index time/op  delta
Reactions/postgres/GetForPost-8   685µs ± 1%   732µs ± 4%  +6.92%  (p=0.000 n=8+10)
                       base time/op  with_added_index time/op  delta
Reactions/mysql/GetForPost-8      753µs ± 3%   805µs ± 3%  +6.97%  (p=0.000 n=10+10)
```

Reply:
Amazing work @streamer45 ! I'll join in with the recommendation of handling the UPDATE in the application logic itself via COALESCE.

And even the column additions are in the order of seconds. So I'll suggest to go the route of offline SQL queries - what we did when we changed the multi-column index for this table.

Essentially, have a very EXPLICIT mention in the release notes to run the upgrade query offline during low traffic hours. And _then_ upgrade the server. Otherwise, the server will be blocked for a while.

Reply:
I've removed the code to clear out the NULLs on database upgrade, and updated all the SELECTS to use COALESCE via https://github.com/mattermost/mattermost-server/pull/16654/commits/8a66a4dac02d59091e6b4cab09969574443ccf7d